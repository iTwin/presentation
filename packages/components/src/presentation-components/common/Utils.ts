/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Core
 */

import { LegacyRef, MutableRefObject, RefCallback, useCallback, useRef, useState } from "react";
import { Primitives, PrimitiveValue, PropertyDescription, PropertyRecord, PropertyValueFormat } from "@itwin/appui-abstract";
import { IPropertyValueRenderer, PropertyValueRendererManager } from "@itwin/components-react";
import { assert, Guid, GuidString, IDisposable } from "@itwin/core-bentley";
import { Descriptor, Field, LabelCompositeValue, LabelDefinition, parseCombinedFieldNames } from "@itwin/presentation-common";
import { Presentation } from "@itwin/presentation-frontend";
import { InstanceKeyValueRenderer } from "../properties/InstanceKeyValueRenderer";

const localizationNamespaceName = "PresentationComponents";

/**
 * Registers 'PresentationComponents' localization namespace and returns callback
 * to unregister it.
 * @internal
 */
export const initializeLocalization = async () => {
  await Presentation.localization.registerNamespace(localizationNamespaceName);
  return () => Presentation.localization.unregisterNamespace(localizationNamespaceName); // eslint-disable-line @itwin/no-internal
};

/**
 * Registers custom property value renderers and returns cleanup callback that unregisters them.
 * @internal
 */
export const initializePropertyValueRenderers = async () => {
  const customRenderers: Array<{ name: string; renderer: IPropertyValueRenderer }> = [{ name: "SelectableInstance", renderer: new InstanceKeyValueRenderer() }];

  for (const { name, renderer } of customRenderers) {
    PropertyValueRendererManager.defaultManager.registerRenderer(name, renderer);
  }

  return () => {
    for (const { name } of customRenderers) {
      PropertyValueRendererManager.defaultManager.unregisterRenderer(name);
    }
  };
};

/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = (stringId: string): string => {
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.localization.getLocalizedString(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName) {
    return component.displayName;
  }
  if (component.name) {
    return component.name;
  }
  return "Component";
};

/**
 * Finds a field given the name of property record created from that field.
 * @internal
 */
export const findField = (descriptor: Descriptor, recordPropertyName: string): Field | undefined => {
  let fieldsSource: { getFieldByName: (name: string) => Field | undefined } | undefined = descriptor;
  const fieldNames = parseCombinedFieldNames(recordPropertyName);
  while (fieldsSource && fieldNames.length) {
    const field: Field | undefined = fieldsSource.getFieldByName(fieldNames.shift()!);
    fieldsSource = field && field.isNestedContentField() ? field : undefined;
    if (!fieldNames.length) {
      return field;
    }
  }
  return undefined;
};

/**
 * Creates property record for label using label definition.
 * @internal
 */
export const createLabelRecord = (label: LabelDefinition, name: string): PropertyRecord => {
  const value: PrimitiveValue = {
    displayValue: label.displayValue,
    value: createPrimitiveLabelValue(label),
    valueFormat: PropertyValueFormat.Primitive,
  };
  const property: PropertyDescription = {
    displayLabel: "Label",
    typename: label.typeName,
    name,
  };

  return new PropertyRecord(value, property);
};

const createPrimitiveLabelValue = (label: LabelDefinition) => {
  return LabelDefinition.isCompositeDefinition(label) ? createPrimitiveCompositeValue(label.rawValue) : label.rawValue;
};

const createPrimitiveCompositeValue = (compositeValue: LabelCompositeValue): Primitives.Composite => {
  return {
    separator: compositeValue.separator,
    parts: compositeValue.values.map((part) => ({
      displayValue: part.displayValue,
      typeName: part.typeName,
      rawValue: createPrimitiveLabelValue(part),
    })),
  };
};

/**
 * A helper to track ongoing async tasks. Usage:
 * ```
 * await using(tracker.trackAsyncTask(), async (_r) => {
 *   await doSomethingAsync();
 * });
 * ```
 *
 * Can be used with `waitForPendingAsyncs` in test helpers to wait for all
 * async tasks to complete.
 *
 * @internal
 */
export class AsyncTasksTracker {
  private _asyncsInProgress = new Set<GuidString>();
  public get pendingAsyncs() {
    return this._asyncsInProgress;
  }
  public trackAsyncTask(): IDisposable {
    const id = Guid.createValue();
    this._asyncsInProgress.add(id);
    return {
      dispose: () => this._asyncsInProgress.delete(id),
    };
  }
}

/** @internal */
export function useResizeObserver<T extends HTMLElement>() {
  const observer = useRef<ResizeObserver>();
  const [{ width, height }, setSize] = useState<{ width?: number; height?: number }>({});

  const ref = useCallback((element: T | null) => {
    observer.current?.disconnect();
    if (element) {
      observer.current = new ResizeObserver(
        /* istanbul ignore next */
        (entries) => {
          assert(entries.length === 1);
          setSize(entries[0].contentRect);
        }
      );
      observer.current.observe(element);
    }
  }, []);

  return {
    ref,
    width,
    height,
  };
}

/** @internal */
export function mergeRefs<T>(...refs: Array<MutableRefObject<T | null> | LegacyRef<T>>): RefCallback<T> {
  return (instance: T | null) => {
    refs.forEach((ref) => {
      // istanbul ignore else
      if (typeof ref === "function") {
        ref(instance);
      } else if (ref) {
        (ref as MutableRefObject<T | null>).current = instance;
      }
    });
  };
}

/**
 * A hook that helps components throw errors in React's render loop so they can be captured by React error
 * boundaries.
 *
 * Usage: simply call the returned function with an error and it will be re-thrown on next render.
 *
 * @internal
 */
export function useErrorState() {
  const [error, setError] = useState<Error | undefined>(undefined);
  const setErrorState = useCallback((e: unknown) => {
    setError(e instanceof Error ? e : /* istanbul ignore next */ new Error());
  }, []);
  if (error) {
    throw error;
  }
  return setErrorState;
}
