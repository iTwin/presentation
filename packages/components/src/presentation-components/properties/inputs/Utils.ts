/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { Format, FormatType, Parser, ParserSpec } from "@itwin/core-quantity";

const FRACTIONAL_PRECISION = "1/16";

/**
 * Finds rounding error for entered value and converts it to persistence unit.
 * @internal
 */
export function getPersistenceUnitRoundingError(numberStr: string, parser: ParserSpec): number | undefined {
  const tokens = Parser.parseQuantitySpecification(numberStr, parser.format);
  const enteredUnit = tokens.length > 0 && tokens[tokens.length - 1].isString ? (tokens[tokens.length - 1].value as string) : undefined;

  // find unit of entered value that will be used when determining precision
  const precisionStr = getPrecision(numberStr, parser.format);
  if (!precisionStr) {
    return undefined;
  }

  // convert precision to persistence unit
  const parseResult = parser.parseToQuantityValue(enteredUnit ? `${precisionStr}${enteredUnit}` : precisionStr);
  return parseResult.ok ? parseResult.value : undefined;
}

/** @internal */
export function getDecimalRoundingError(numStr: string) {
  const precision = getDecimalPrecision(numStr);
  return precision ? Number(precision) : undefined;
}

function getPrecision(numberStr: string, format: Format) {
  // use decimal precision if number contains decimal separator
  if (numberStr.includes(getLocalizedDecimalSeparator())) {
    return getDecimalPrecision(numberStr);
  }

  // use fractional precision if number contains fraction separator
  if (numberStr.includes("/")) {
    return FRACTIONAL_PRECISION;
  }

  // if number does not have decimal or fraction separator use precision based on format type.
  // TODO: this might not be correct in all cases. For example:
  // if format is `Fractional` but entered value is `2 m` precision should be `0.5 m` but we will get ` 1/16 m`
  if (format.type === FormatType.Decimal) {
    return getDecimalPrecision(numberStr);
  }

  if (format.type === FormatType.Fractional) {
    return FRACTIONAL_PRECISION;
  }

  return undefined;
}

let localeSpecificDecimalSeparator: string | undefined;
function getLocalizedDecimalSeparator(): string {
  if (localeSpecificDecimalSeparator !== undefined) {
    return localeSpecificDecimalSeparator;
  }

  localeSpecificDecimalSeparator = ".";
  const matches = (12345.6789).toLocaleString().match(/345(.*)67/);
  if (matches && matches.length > 1) {
    localeSpecificDecimalSeparator = matches[1];
  }

  return localeSpecificDecimalSeparator;
}

function getDecimalPrecision(numStr: string): string | undefined {
  const separator = getLocalizedDecimalSeparator();
  let lastDigit = -1;
  // find last digit of the number
  for (let i = numStr.length - 1; i >= 0; i--) {
    if (isDigit(numStr[i])) {
      lastDigit = i;
      break;
    }
  }

  if (lastDigit === -1) {
    return undefined;
  }

  let separatorIndex = -1;
  // find separator position in the number
  for (let i = lastDigit - 1; i >= 0; i--) {
    if (numStr[i] === separator) {
      separatorIndex = i;
      break;
    }

    if (!isDigit(numStr[i])) {
      break;
    }
  }

  if (separatorIndex === -1) {
    return "0.5";
  }

  const digitsAfterSeparator = lastDigit - separatorIndex;
  return `${0.5 * Math.pow(10, -digitsAfterSeparator)}`;
}

function isDigit(char: string) {
  return Number.isFinite(parseInt(char, 10));
}
