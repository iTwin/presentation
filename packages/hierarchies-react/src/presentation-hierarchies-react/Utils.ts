/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { useCallback, useInsertionEffect, useRef } from "react";

/** @internal */
export function useLatest<T>(value: T) {
  const ref = useRef(value);

  useInsertionEffect(() => {
    ref.current = value;
  }, [value]);

  return ref;
}

/** @internal */
export function useEvent<TCallback extends (...args: any[]) => void>(callback: TCallback) {
  const latestCallback = useLatest(callback);

  return useCallback(
    (...args: Parameters<TCallback>) => {
      latestCallback.current(...args);
    },
    [latestCallback],
  );
}
