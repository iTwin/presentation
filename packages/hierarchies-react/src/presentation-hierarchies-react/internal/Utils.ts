/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { HierarchyNode, HierarchyNodeKey } from "@itwin/presentation-hierarchies";

export function createNodeId(node: Pick<HierarchyNode, "key" | "parentKeys">) {
  return [...node.parentKeys.map(serializeNodeKey), serializeNodeKey(node.key)].join(";");
}

function serializeNodeKey(key: HierarchyNodeKey): string {
  return HierarchyNodeKey.isCustom(key) ? key : convertObjectValuesToString(key);
}

function convertObjectValuesToString(obj: object): string {
  return Object.entries(obj)
    .map(([, value]) => {
      if (typeof value === "object") {
        return convertObjectValuesToString(value);
      }
      return String(value);
    })
    .join(",");
}

export function sameNodes(lhs: HierarchyNode, rhs: HierarchyNode): boolean {
  if (HierarchyNodeKey.compare(lhs.key, rhs.key) !== 0) {
    return false;
  }

  if (lhs.parentKeys.length !== rhs.parentKeys.length) {
    return false;
  }

  for (let i = lhs.parentKeys.length - 1; i >= 0; --i) {
    if (HierarchyNodeKey.compare(lhs.parentKeys[i], rhs.parentKeys[i]) !== 0) {
      return false;
    }
  }
  return true;
}
