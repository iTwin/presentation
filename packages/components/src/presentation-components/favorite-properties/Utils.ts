/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { CategoryDescription } from "@itwin/presentation-common";
import { translate } from "../common/Utils";

/** @internal */
export const FAVORITES_CATEGORY_NAME = "Favorite";

/** @internal */
export const getFavoritesCategory = (): CategoryDescription => {
  return {
    name: FAVORITES_CATEGORY_NAME,
    label: translate("categories.favorite.label"),
    description: translate("categories.favorite.description"),
    priority: Number.MAX_VALUE,
    expand: true,
  };
};
