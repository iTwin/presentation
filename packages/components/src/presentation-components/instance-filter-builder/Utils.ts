/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module InstancesFilter
 */

import { useMemo } from "react";
import { PrimitiveValue, PropertyDescription, PropertyValueFormat, StandardTypeNames } from "@itwin/appui-abstract";
import {
  defaultPropertyFilterBuilderRuleValidator,
  isUnaryPropertyFilterOperator,
  PropertyFilterBuilderRule,
  PropertyFilterRuleOperator,
} from "@itwin/components-react";
import { IModelConnection } from "@itwin/core-frontend";
import { CategoryDescription, ClassInfo, combineFieldNames, Descriptor, Field, NestedContentField, PropertiesField } from "@itwin/presentation-common";
import { createPropertyDescriptionFromFieldInfo } from "../common/ContentBuilder";
import { translate } from "../common/Utils";
import { NavigationPropertyEditorContextProps } from "../properties/editors/NavigationPropertyEditorContext";
import { PresentationInstanceFilterPropertyInfo } from "./PresentationFilterBuilder";

/** @internal */
export function createInstanceFilterPropertyInfos(descriptor: Descriptor): PresentationInstanceFilterPropertyInfo[] {
  const propertyInfos = new Array<PresentationInstanceFilterPropertyInfo>();
  for (const field of descriptor.fields) {
    propertyInfos.push(...createPropertyInfos(field));
  }
  return propertyInfos;
}

/** @internal */
export function getInstanceFilterFieldName(property: PropertyDescription) {
  const [_, fieldName] = property.name.split(INSTANCE_FILTER_FIELD_SEPARATOR);
  return fieldName;
}

/** @internal */
export const DEFAULT_ROOT_CATEGORY_NAME = "/selected-item/";

function getPropertySourceClassInfo(field: PropertiesField | NestedContentField): ClassInfo {
  if (field.parent) {
    return getPropertySourceClassInfo(field.parent);
  }

  if (field.isPropertiesField()) {
    return field.properties[0].property.classInfo;
  }
  return field.pathToPrimaryClass[field.pathToPrimaryClass.length - 1].targetClassInfo;
}

function getPropertyClassInfo(field: PropertiesField): ClassInfo {
  if (field.parent && field.parent.isNestedContentField()) {
    return field.parent.contentClassInfo;
  }

  return field.properties[0].property.classInfo;
}

function createPropertyInfos(field: Field): PresentationInstanceFilterPropertyInfo[] {
  if (field.isNestedContentField()) {
    return field.nestedFields.flatMap((nestedField) => createPropertyInfos(nestedField));
  }
  // istanbul ignore else
  if (field.isPropertiesField()) {
    return [createPropertyInfoFromPropertiesField(field)];
  }
  // istanbul ignore next
  return [];
}

interface CategoryInfo {
  name?: string;
  label?: string;
}

function getCategoryInfo(category: CategoryDescription, categoryInfo: CategoryInfo): CategoryInfo {
  if (category.name === DEFAULT_ROOT_CATEGORY_NAME) {
    return categoryInfo;
  }

  const newInfo = {
    name: categoryInfo.name ? `${category.name}/${categoryInfo.name}` : `${category.name}`,
    label: categoryInfo.label ? `${category.label} | ${categoryInfo.label}` : `${category.label}`,
  };

  return category.parent ? getCategoryInfo(category.parent, newInfo) : newInfo;
}

function getParentNames(field: Field, name: string): string {
  if (!field.parent) {
    return combineFieldNames(name, field.name);
  }
  return getParentNames(field.parent, combineFieldNames(name, field.name));
}

/** @internal */
export function createPropertyInfoFromPropertiesField(field: PropertiesField): PresentationInstanceFilterPropertyInfo {
  const categoryInfo = getCategoryInfo(field.category, { name: undefined, label: undefined });
  const name = field.parent ? getParentNames(field.parent, field.name) : field.name;

  const propertyDescription = createPropertyDescriptionFromFieldInfo({
    name: getCategorizedFieldName(name, categoryInfo.name),
    label: field.label,
    type: field.type,
    editor: field.editor,
    enum: field.properties[0].property.enumerationInfo,
    isReadonly: field.isReadonly,
    renderer: field.renderer,
    koqName: field.properties[0].property.kindOfQuantity?.name,
  });

  return {
    field,
    sourceClassId: getPropertySourceClassInfo(field).id,
    propertyDescription,
    categoryLabel: categoryInfo.label,
    className: getPropertyClassInfo(field).name,
  };
}

/** @internal */
export const INSTANCE_FILTER_FIELD_SEPARATOR = "#";
function getCategorizedFieldName(fieldName: string, categoryName?: string) {
  return `${categoryName ?? ""}${INSTANCE_FILTER_FIELD_SEPARATOR}${fieldName}`;
}

/** @internal */
export function useFilterBuilderNavigationPropertyEditorContext(imodel: IModelConnection, descriptor: Descriptor) {
  return useMemo<NavigationPropertyEditorContextProps>(
    () => ({
      imodel,
      getNavigationPropertyInfo: async (property) => {
        const field = descriptor.getFieldByName(getInstanceFilterFieldName(property));
        if (!field || !field.isPropertiesField()) {
          return undefined;
        }

        return field.properties[0].property.navigationPropertyInfo;
      },
    }),
    [imodel, descriptor],
  );
}

/** @internal */
export function filterRuleValidator(item: PropertyFilterBuilderRule) {
  // skip empty rules and rules that do not require value
  if (item.property === undefined || item.operator === undefined || isUnaryPropertyFilterOperator(item.operator)) {
    return undefined;
  }

  // istanbul ignore if
  if (item.value !== undefined && item.value.valueFormat !== PropertyValueFormat.Primitive) {
    return undefined;
  }

  const error = combineValidators(
    quantityPropertyValidator,
    numericPropertyValidator,
  )({
    property: item.property,
    operator: item.operator,
    value: item.value,
  });

  if (error) {
    return error;
  }

  return defaultPropertyFilterBuilderRuleValidator(item);
}

function combineValidators(...validators: Array<(ctx: ValidatorContext) => string | undefined>) {
  return (ctx: ValidatorContext) => {
    for (const validator of validators) {
      const error = validator(ctx);
      if (error) {
        return error;
      }
    }
    return undefined;
  };
}

interface ValidatorContext {
  property: PropertyDescription;
  operator: PropertyFilterRuleOperator;
  value?: PrimitiveValue;
}

function quantityPropertyValidator({ property, value }: ValidatorContext) {
  // rules with non quantity properties or without values
  if (property.quantityType === undefined || value === undefined) {
    return undefined;
  }
  if (isInvalidNumericValue(value)) {
    return translate("instance-filter-builder.error-messages.invalid");
  }

  return undefined;
}

function numericPropertyValidator({ property, value }: ValidatorContext) {
  if (!isPropertyNumeric(property.typename) || value === undefined) {
    return undefined;
  }

  if (isInvalidNumericValue(value)) {
    return translate("instance-filter-builder.error-messages.not-a-number");
  }

  return undefined;
}

function isPropertyNumeric(typename: string) {
  return (
    typename === StandardTypeNames.Number || typename === StandardTypeNames.Int || typename === StandardTypeNames.Float || typename === StandardTypeNames.Double
  );
}

function isInvalidNumericValue(value: PrimitiveValue) {
  return value.displayValue !== undefined && value.displayValue !== "" && isNaN(Number(value.value));
}
