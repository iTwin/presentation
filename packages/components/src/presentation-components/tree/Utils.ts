/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import { Observable as RxjsObservable } from "rxjs/internal/Observable";
import { PropertyRecord } from "@itwin/appui-abstract";
import { DelayLoadedTreeNodeItem, ItemColorOverrides, ItemStyle, Observable, TreeNodeItem, PageOptions as UiPageOptions } from "@itwin/components-react";
import { CheckBoxState } from "@itwin/core-react";
import { LabelDefinition, Node, NodeKey, PartialNode, PageOptions as PresentationPageOptions } from "@itwin/presentation-common";
import { StyleHelper } from "../common/StyleHelper";
import { createLabelRecord } from "../common/Utils";
import { InfoTreeNodeItemType, PresentationInfoTreeNodeItem, PresentationTreeNodeItem } from "./PresentationTreeNodeItem";

/** @internal */
export interface CreateTreeNodeItemProps {
  appendChildrenCountForGroupingNodes?: boolean;
  customizeTreeNodeItem?: (item: Partial<DelayLoadedTreeNodeItem>, node: Partial<Node>) => void;
}

/** @internal */
export function createTreeNodeItems(nodes: ReadonlyArray<Readonly<Node>>, parentId?: string, props?: CreateTreeNodeItemProps): PresentationTreeNodeItem[] {
  const list = new Array<PresentationTreeNodeItem>();
  for (const node of nodes) {
    list.push(createTreeNodeItem(node, parentId, props));
  }
  return list;
}

/** @internal */
export function createTreeNodeItem(node: Readonly<Node>, parentId?: string, props?: CreateTreeNodeItemProps): PresentationTreeNodeItem {
  const item: PresentationTreeNodeItem = {
    id: createTreeNodeId(node.key),
    label: createNodeLabelRecord(node, !!props?.appendChildrenCountForGroupingNodes),
    key: node.key,
  };
  assignOptionalTreeNodeItemFields(item, node, parentId);
  const customizeItemCallback = props?.customizeTreeNodeItem ?? customizeTreeNodeItem; // eslint-disable-line deprecation/deprecation
  customizeItemCallback(item, node);
  return item;
}

/** @internal */
export function createPartialTreeNodeItem(node: PartialNode, parentId: string | undefined, props: CreateTreeNodeItemProps): Partial<PresentationTreeNodeItem> {
  const item: Partial<PresentationTreeNodeItem> = {};
  if (node.key !== undefined) {
    item.id = createTreeNodeId(node.key);
    item.label = createNodeLabelRecord(node, !!props.appendChildrenCountForGroupingNodes);
  }

  assignOptionalTreeNodeItemFields(item, node, parentId);
  const customizeItemCallback = props.customizeTreeNodeItem ?? customizeTreeNodeItem; // eslint-disable-line deprecation/deprecation
  customizeItemCallback(item, node);
  return item;
}

/** @internal */
export function createTreeNodeId(key: NodeKey): string {
  return [...key.pathFromRoot].reverse().join("/");
}

function assignOptionalTreeNodeItemFields(item: Partial<PresentationTreeNodeItem>, node: Partial<Node>, parentId?: string): void {
  if (node.key !== undefined) {
    item.key = node.key;
  }

  if (parentId) {
    item.parentId = parentId;
  }

  if (node.description) {
    item.description = node.description;
  }

  if (node.hasChildren) {
    item.hasChildren = true;
  }

  if (node.isExpanded) {
    item.autoExpand = true;
  }

  if (node.extendedData) {
    item.extendedData = node.extendedData;
  }
}

/**
 * Applies customization from [Node]($presentation-common) to [TreeNodeItem]($components-react).
 * @public
 * @deprecated in 3.x. This is a temporary function to apply deprecated [Node]($presentation-common) customization attributes on a [TreeNodeItem]($components-react).
 * The recommendation is to switch to using a custom customization function that looks at node's [Node.extendedData]($presentation-common) and applies
 * [TreeNodeItem]($components-react) customization based on values in extended data.
 * See [extended data usage page]($docs/presentation/customization/ExtendedDataUsage.md) for more details.
 */
// istanbul ignore next
export function customizeTreeNodeItem(item: Partial<DelayLoadedTreeNodeItem>, node: Partial<Node>) {
  // eslint-disable-next-line deprecation/deprecation
  if (node.imageId) {
    item.icon = node.imageId; // eslint-disable-line deprecation/deprecation
  }

  // eslint-disable-next-line deprecation/deprecation
  if (node.isCheckboxVisible) {
    item.isCheckboxVisible = true;
    // eslint-disable-next-line deprecation/deprecation
    if (node.isChecked) {
      // eslint-disable-line deprecation/deprecation
      item.checkBoxState = CheckBoxState.On;
    }

    // eslint-disable-next-line deprecation/deprecation
    if (!node.isCheckboxEnabled) {
      item.isCheckboxDisabled = true;
    }
  }

  const style = createTreeNodeItemStyle(node);
  if (Object.keys(style).length > 0) {
    item.style = style;
  }
}

// istanbul ignore next
function createTreeNodeItemStyle(node: Partial<Node>): ItemStyle {
  const style: ItemStyle = {};
  if (StyleHelper.isBold(node)) {
    style.isBold = true;
  }

  if (StyleHelper.isItalic(node)) {
    style.isItalic = true;
  }

  const colorOverrides: ItemColorOverrides = {};
  const foreColor = StyleHelper.getForeColor(node);
  if (foreColor) {
    colorOverrides.color = foreColor;
  }

  const backColor = StyleHelper.getBackColor(node);
  if (backColor) {
    colorOverrides.backgroundColor = backColor;
  }

  if (Object.keys(colorOverrides).length > 0) {
    style.colorOverrides = colorOverrides;
  }

  return style;
}

/** @internal */
export function pageOptionsUiToPresentation(pageOptions?: UiPageOptions): PresentationPageOptions | undefined {
  if (pageOptions) {
    return { ...pageOptions };
  }
  return undefined;
}

function createNodeLabelRecord(node: Node, appendChildrenCountForGroupingNodes: boolean): PropertyRecord {
  let labelDefinition = node.label;
  if (appendChildrenCountForGroupingNodes && NodeKey.isGroupingNodeKey(node.key)) {
    const countDefinition: LabelDefinition = {
      displayValue: `(${node.key.groupedInstancesCount})`,
      rawValue: `(${node.key.groupedInstancesCount})`,
      typeName: "string",
    };
    labelDefinition = {
      displayValue: `${labelDefinition.displayValue} ${countDefinition.displayValue}`,
      rawValue: {
        separator: " ",
        values: [labelDefinition, countDefinition],
      },
      typeName: LabelDefinition.COMPOSITE_DEFINITION_TYPENAME,
    };
  }
  return createLabelRecord(labelDefinition, "node_label");
}

/** @internal */
export function toRxjsObservable<T>(source: Observable<T>): RxjsObservable<T> {
  return new RxjsObservable((subscriber) => source.subscribe(subscriber));
}

export function createInfoNode(parentNode: TreeNodeItem | undefined, message: string, type?: InfoTreeNodeItemType): PresentationInfoTreeNodeItem {
  const id = parentNode ? `${parentNode.id}/info-node` : `/info-node/${message}`;
  return {
    id,
    parentId: parentNode?.id,
    label: PropertyRecord.fromString(message),
    message,
    isSelectionDisabled: true,
    children: undefined,
    type: type ?? InfoTreeNodeItemType.Unset,
  };
}
