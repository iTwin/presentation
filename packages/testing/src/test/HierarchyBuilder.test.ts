/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { expect } from "chai";
import sinon from "sinon";
import { TreeNodeItem } from "@itwin/components-react";
import { BeEvent, Guid } from "@itwin/core-bentley";
import { IModelConnection } from "@itwin/core-frontend";
import { HierarchyRequestOptions, LabelDefinition, Node, NodeKey, RegisteredRuleset, Ruleset } from "@itwin/presentation-common";
import { Presentation, PresentationManager, RulesetManager, RulesetVariablesManager } from "@itwin/presentation-frontend";
import { HierarchyBuilder, NodeMappingFunc } from "../presentation-testing/HierarchyBuilder";
import { createStub } from "./Utils";

async function getRootNodes() {
  const root: Node = {
    label: LabelDefinition.fromLabelString("Root Node"),
    hasChildren: true,
    key: { type: "", version: 0, pathFromRoot: ["root"] },
  };
  return { nodes: [root], count: 1 };
}

async function getChildrenNodes(opts: HierarchyRequestOptions<IModelConnection, NodeKey>) {
  if (opts.parentKey?.pathFromRoot[0] !== "root" || opts?.parentKey.pathFromRoot.length !== 1) {
    return { nodes: [], count: 0 };
  }

  const child1: Node = {
    label: LabelDefinition.fromLabelString("Child 1"),
    key: { type: "", version: 0, pathFromRoot: ["root", "child1"] },
  };
  const child2: Node = {
    label: LabelDefinition.fromLabelString("Child 2"),
    key: { type: "", version: 0, pathFromRoot: ["root", "child2"] },
  };
  return { nodes: [child1, child2], count: 2 };
}

describe("HierarchyBuilder", () => {
  let presentationManager: sinon.SinonStubbedInstance<PresentationManager>;
  const rulesetManager = {
    add: createStub<RulesetManager["add"]>(),
  };

  const ruleset = { id: "1" } as Ruleset;
  const imodel = {} as IModelConnection;

  beforeEach(() => {
    rulesetManager.add.callsFake(async (rules) => new RegisteredRuleset(rules, Guid.createValue(), () => {}));

    presentationManager = sinon.createStubInstance(PresentationManager);
    presentationManager.rulesets.returns(rulesetManager as unknown as RulesetManager);
    presentationManager.vars.returns({
      onVariableChanged: new BeEvent(),
    } as RulesetVariablesManager);
  });

  afterEach(() => {
    rulesetManager.add.reset();
  });

  describe("createHierarchy", () => {
    context("without data", () => {
      beforeEach(() => {
        sinon.stub(Presentation, "presentation").get(() => presentationManager);
        presentationManager.getNodesAndCount.resolves({ nodes: [], count: 0 });
      });

      afterEach(() => {
        sinon.restore();
      });

      it("returns empty list when rulesetId is given", async () => {
        const builder = new HierarchyBuilder({ imodel });
        const hierarchy = await builder.createHierarchy("1");
        expect(hierarchy).to.be.empty;
      });

      it("returns empty list when ruleset is given", async () => {
        const builder = new HierarchyBuilder({ imodel });
        const hierarchy = await builder.createHierarchy(ruleset);
        expect(hierarchy).to.be.empty;
      });
    });

    context("with data", () => {
      beforeEach(() => {
        sinon.stub(Presentation, "presentation").get(() => presentationManager);
        presentationManager.getNodesAndCount.callsFake(async (opts) => (opts.parentKey === undefined ? getRootNodes() : getChildrenNodes(opts)));
      });

      afterEach(() => {
        sinon.restore();
      });

      it("returns correct hierarchy", async () => {
        const builder = new HierarchyBuilder({ imodel });
        expect(await builder.createHierarchy(ruleset)).to.matchSnapshot();
      });

      it("returns correct hierarchy with custom node mapping function", async () => {
        const nodeMapper: NodeMappingFunc = (node: TreeNodeItem) => ({ id: node.id });
        const builder = new HierarchyBuilder({ imodel, nodeMappingFunc: nodeMapper });
        expect(await builder.createHierarchy(ruleset)).to.matchSnapshot();
      });
    });
  });
});
