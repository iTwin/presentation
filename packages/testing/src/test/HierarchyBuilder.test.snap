// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchyBuilder createHierarchy with data returns correct hierarchy 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "label": PropertyRecord {
          "property": Object {
            "displayLabel": "Label",
            "name": "node_label",
            "typename": "string",
          },
          "value": Object {
            "displayValue": "Child 1",
            "value": "Child 1",
            "valueFormat": 0,
          },
        },
      },
      Object {
        "label": PropertyRecord {
          "property": Object {
            "displayLabel": "Label",
            "name": "node_label",
            "typename": "string",
          },
          "value": Object {
            "displayValue": "Child 2",
            "value": "Child 2",
            "valueFormat": 0,
          },
        },
      },
    ],
    "hasChildren": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Root Node",
        "value": "Root Node",
        "valueFormat": 0,
      },
    },
  },
]
`;

exports[`HierarchyBuilder createHierarchy with data returns correct hierarchy with custom node mapping function 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "id": "child1/root",
      },
      Object {
        "id": "child2/root",
      },
    ],
    "id": "root",
  },
]
`;
